From 25120743e07e475d2f5ca5a17ae28b58fc1c4647 Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Tue, 15 Dec 2020 11:27:05 +0800
Subject: [PATCH] drm/rockchip: vop2: fix esmart display black screen

This is workaround solution for IC design:
esmart can't support scale down when actual_w % 16 == 1.

Change-Id: I0fa090c5e28b79e644ef31ea1184b724b9d32ffa
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index ac19a27abe93..4a726d2fdc58 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -1946,6 +1946,15 @@ static void vop2_plane_atomic_update(struct drm_plane *plane, struct drm_plane_s
 			  win->name, dest->y1, dsp_h, adjusted_mode->vdisplay);
 		dsp_h = adjusted_mode->vdisplay - dest->y1;
 	}
+
+	/*
+	 * This is workaround solution for IC design:
+	 * esmart can't support scale down when actual_w % 16 == 1.
+	 */
+	if (!(win->feature & WIN_FEATURE_AFBDC) &&
+	    actual_w > dsp_w && actual_w % 16 == 1)
+		actual_w -= 1;
+
 	act_info = (actual_h - 1) << 16 | ((actual_w - 1) & 0xffff);
 	dsp_info = (dsp_h - 1) << 16 | ((dsp_w - 1) & 0xffff);
 	stride = DIV_ROUND_UP(fb->pitches[0], 4);
-- 
2.35.3

